litedram/core/bankmachine.py

```diff
@@ -45,73 +49,118 @@ def __init__(self, n, aw, address_align, nranks, settings):
+            wr_conflict.eq((cmd_buffer_lookahead.source.addr == cmd_bufferRead.source.addr) & cmd_bufferRead.source.valid),
+
             req.connect(cmd_buffer_lookahead.sink, keep={"valid", "ready", "we", "addr"}),
-            cmd_buffer_lookahead.source.connect(cmd_buffer.sink),
-            cmd_buffer.source.ready.eq(req.wdata_ready | req.rdata_valid),
-            req.lock.eq(cmd_buffer_lookahead.source.valid | cmd_buffer.source.valid),
-        ]
+            
+            cmd_buffer_lookahead.source.connect(cmd_bufferRead.sink, omit={"valid", "ready"}),
+            cmd_buffer_lookahead.source.connect(cmd_bufferWrite.sink, omit={"valid", "ready"}),
+
+            cmd_bufferRead.sink.valid.eq(
+                cmd_buffer_lookahead.source.valid & ~cmd_buffer_lookahead.source.we & ~rd_conflict
+            ),
+            cmd_bufferWrite.sink.valid.eq(
+                cmd_buffer_lookahead.source.valid & cmd_buffer_lookahead.source.we & ~wr_conflict
+            ),
+
+            cmd_buffer_lookahead.source.ready.eq(
+                ((cmd_bufferRead.sink.ready | cmd_bufferRead.source.ready) & ~cmd_buffer_lookahead.source.we & ~rd_conflict)
+                | ((cmd_bufferWrite.sink.ready | cmd_bufferWrite.source.ready) & cmd_buffer_lookahead.source.we & ~wr_conflict)
+            ),
+
+            cmd_bufferRead.source.ready.eq(req.rdata_valid),
+            cmd_bufferWrite.source.ready.eq(req.wdata_ready),
 
+            req.lock.eq(cmd_buffer_lookahead.source.valid | cmd_bufferRead.source.valid | cmd_bufferWrite.source.valid),
+        ]
+        
+        cmd_buffer = Record(cmd_buffer_layout + [("valid",1)])
         slicer = _AddressSlicer(settings.geom.colbits, address_align)
 
+        
         # Row tracking
         has_openrow = Signal()
         openrow = Signal(settings.geom.rowbits, reset_less=True)
         hit = Signal()
-        self.comb += hit.eq(openrow == slicer.row(cmd_buffer.source.addr))
         track_open = Signal()
         track_close = Signal()
         self.sync += \
             If(track_close,
                 has_openrow.eq(0)
             ).Elif(track_open,
                 has_openrow.eq(1),
-                openrow.eq(slicer.row(cmd_buffer.source.addr))
+                openrow.eq(slicer.row(cmd_buffer.addr))
             )
 
+        # Chose the correct source for the command
+        sync_or_comb = None
+        if settings.timing.tCCD > 1:
+            # This inserts a pipeline bubble of one cycle and so cannot be used when a command is required every cycle
+            sync_or_comb = self.sync
+        else:
+            sync_or_comb = self.comb
+        sync_or_comb += [
+            If((self.want_writes & cmd_bufferWrite.source.valid) | ~cmd_bufferRead.source.valid,
+                # Note: when sync valid is asserted one extra cycle (OK because tCCD prevents issuing a duplicate command)
```

Comment:
It's even more subtle than this.  If cmd.valid were deasserted in the bubble cycle then the arbitrator in the multiplixer would move on.  This would require an extra cycle for recovery limiting sequential throughput to 2/3.